test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java

```diff
@@ -19,97 +19,76 @@
+        // TODO: Decide when to start adding more bytes. We do not actually want to wait until buffer is exhausted.
+        if (channelBuffer.hasRemaining() == false) {
+            channelBuffer.addBuffer(allocator.get());
         }
 
-        int bytesRead = channel.read(references.getLast());
+        int bytesRead = channel.read(channelBuffer);
 
         if (bytesRead == -1) {
             return bytesRead;
         }
 
-        rawBytesCount += bytesRead;
-
-        BytesReference message;
+        ChannelMessage message;
 
         // Frame decoder will throw an exception if the message is improperly formatted, the header is incorrect,
         // or the message is corrupted
-        while ((message = frameDecoder.decode(createCompositeBuffer(), rawBytesCount)) != null) {
-            int messageLengthWithHeader = message.length();
-            NetworkBytesReference.vectorizedIncrementReadIndexes(references, messageLengthWithHeader);
-            trimDecodedMessages(messageLengthWithHeader);
-            rawBytesCount -= messageLengthWithHeader;
+        while ((message = frameDecoder.decode(channelBuffer)) != null) {
+            BytesReference messageBytes = message.getContent();
+            int messageLengthWithHeader = messageBytes.length();
 
             try {
-                BytesReference messageWithoutHeader = message.slice(6, message.length() - 6);
+                BytesReference messageWithoutHeader = messageBytes.slice(6, messageBytes.length() - 6);
 
                 // A message length of 6 bytes it is just a ping. Ignore for now.
                 if (messageLengthWithHeader != 6) {
                     handler.handleMessage(messageWithoutHeader, channel, channel.getProfile(), messageWithoutHeader.length());
                 }
             } catch (Exception e) {
                 handler.handleException(channel, e);
+            } finally {
+                message.close();
             }
         }
 
         return bytesRead;
     }
 
-    private CompositeBytesReference createCompositeBuffer() {
-        return new CompositeBytesReference(references.toArray(new BytesReference[references.size()]));
-    }
-
-    private void trimDecodedMessages(int bytesToTrim) {
-        while (bytesToTrim != 0) {
-            NetworkBytesReference ref = references.getFirst();
-            int readIndex = ref.getReadIndex();
-            bytesToTrim -= readIndex;
-            if (readIndex == ref.length()) {
-                references.removeFirst();
-            } else {
-                assert bytesToTrim == 0;
-                if (readIndex != 0) {
-                    references.removeFirst();
-                    NetworkBytesReference slicedRef = ref.slice(readIndex, ref.length() - readIndex);
-                    references.addFirst(slicedRef);
-                }
-            }
-
+    @Override
+    public void close() {
+        if (closed == false) {
+            channelBuffer.close();
```

Comment:
maybe just protected channelbuffer from having sideeffects of multiple closing and remove the closed flag? it seems unused?